src/PowerShellEditorServices/Workspace/WorkspaceFileSystemWrapper.cs

```diff
@@ -0,0 +1,328 @@
+        {
+            if (!concreteDirectoryInfo.Exists || depth >= factory.MaxRecursionDepth) { yield break; }
+            foreach (var fileSystemInfo in this.factory.SafeEnumerateFileSystemInfos(this.concreteDirectoryInfo))
+            {
+                var directoryInfo = fileSystemInfo as DirectoryInfo;
+                if (directoryInfo != null)
+                {
+                    yield return this.factory.CreateDirectoryInfoWrapper(directoryInfo, depth + 1);
+                }
+                else
+                {
+                    yield return this.factory.CreateFileInfoWrapper((FileInfo)fileSystemInfo, depth);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns an instance of <see cref="DirectoryInfoBase" /> that represents a subdirectory.
+        /// </summary>
+        /// <remarks>
+        /// If <paramref name="name" /> equals '..', this returns the parent directory.
+        /// </remarks>
+        /// <param name="name">The directory name</param>
+        /// <returns>The directory</returns>
+        public override DirectoryInfoBase GetDirectory(string name)
+        {
+            var isParentPath = string.Equals(name, "..", StringComparison.Ordinal);
+
+            if (isParentPath)
+            {
+                return this.ParentDirectory;
+            }
+            else
+            {
+                var dirs = concreteDirectoryInfo.GetDirectories(name);
+
+                if (dirs.Length == 1)
+                {
+                    return this.factory.CreateDirectoryInfoWrapper(dirs[0], depth + 1);
+                }
+                else if (dirs.Length == 0)
+                {
+                    return null;
+                }
+                else
+                {
+                    // This shouldn't happen. The parameter name isn't supposed to contain wild card.
+                    throw new InvalidOperationException(
+                        string.Format("More than one sub directories are found under {0} with name {1}.",
+                            concreteDirectoryInfo.FullName, name));
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public override FileInfoBase GetFile(string name) => this.factory.CreateFileInfoWrapper(new FileInfo(Path.Combine(concreteDirectoryInfo.FullName, name)), depth);
+
+        /// <inheritdoc />
+        public override string Name => isParentPath ? ".." : concreteDirectoryInfo.Name;
+
+        /// <summary>
+        /// Returns the full path to the directory.
+        /// </summary>
+        public override string FullName => concreteDirectoryInfo.FullName;
+
+        /// <summary>
+        /// Returns the parent directory.
+        /// </summary>
+        public override DirectoryInfoBase ParentDirectory
+        {
+            get {
+                try
+                {
+                    return this.factory.CreateDirectoryInfoWrapper(concreteDirectoryInfo.Parent, this.depth - 1);
```

Comment:
I don't see any checks in the code for when depth goes negative. Presumably when you call `ParentDirectory` on the `Root`. Do we need to safe guard against this somewhere?

Reply:
Added a minimum depth of 0 when creating the file and dir objects.